File: partiql-ast/src/main/java/org/partiql/ast/v1/Let.java
package org.partiql.ast.v1;

import lombok.Builder;
import org.jetbrains.annotations.NotNull;
import org.partiql.ast.v1.expr.Expr;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * TODO docs, equals, hashcode
 */
@Builder
public class Let extends AstNode {
    @NotNull
    public final List<Binding> bindings;

    public Let(@NotNull List<Binding> bindings) {
        this.bindings = bindings;
    }

    @NotNull
    @Override
    public Collection<AstNode> children() {
        return new ArrayList<>(bindings);
    }

    @Override
    public <R, C> R accept(@NotNull AstVisitor<R, C> visitor, C ctx) {
        return visitor.visitLet(this, ctx);
    }

    /**
     * TODO docs, equals, hashcode
     */
    @Builder
    public static class Binding extends AstNode {
        @NotNull
        public final Expr expr;

        @NotNull
        public final Identifier asAlias;

        public Binding(@NotNull Expr expr, @NotNull Identifier asAlias) {
            this.expr = expr;
            this.asAlias = asAlias;
        }

        @NotNull
        @Override
        public Collection<AstNode> children() {
            List<AstNode> kids = new ArrayList<>();
            kids.add(expr);
            kids.add(asAlias);
            return kids;
        }

        @Override
        public <R, C> R accept(@NotNull AstVisitor<R, C> visitor, C ctx) {
            return visitor.visitLetBinding(this, ctx);
        }
    }
}

File: partiql-ast/src/main/kotlin/org/partiql/ast/v1/Let.kt
package org.partiql.ast.v1

/**
 * TODO docs, equals, hashcode
 */
public class Let(
    @JvmField
    public var bindings: List<Binding>,
) : AstNode() {
    public override fun children(): Collection<AstNode> {
        val kids = mutableListOf<AstNode?>()
        kids.addAll(bindings)
        return kids.filterNotNull()
    }

    public override fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R = visitor.visitLet(
        this,
        ctx
    )

    /**
     * TODO docs, equals, hashcode
     */
    public class Binding(
        @JvmField
        public var expr: Expr,
        @JvmField
        public var asAlias: Identifier.Symbol,
    ) : AstNode() {
        public override fun children(): Collection<AstNode> {
            val kids = mutableListOf<AstNode?>()
            kids.add(expr)
            kids.add(asAlias)
            return kids.filterNotNull()
        }

        public override fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R =
            visitor.visitLetBinding(this, ctx)
    }
}
